### ISSUE TITLE
Inconsistent behavior when initialize a hash use `of` and normal statement

### ISSUE THREAD

**Issue by zw963** [crystal-lang/crystal#13135]
## Bug Report

Check https://forum.crystal-lang.org/t/double-splatting-a-union-namedtuple-namedtuple-is-not-yet-supported/5442/3?u=zw963 for more details.

Basically, following code works.

```crystal
options = {:foo => true} of Symbol => Bool | String
```

But following code raise a exception.

```crystal
options : Hash(Symbol, Bool | String) = {:foo => true}
```

Do we have any reason for such inconsistent behavior?

Thanks.

**Comment by straight-shoota** [comment]
I see no iconsistency. `{:foo => true}` is a hash literal and has the type `Hash(Symbol, Bool)`. You cannot assign it to a variable which restricted to the type `Hash(Symbol, Bool | String)`. If you want to do that, you can specify the concrete type with the `of` notation as in the first example:
```cr
options : Hash(Symbol, Bool | String) = {:foo => true} of Symbol => Bool | String
```

**Comment by straight-shoota** [comment]
Perhaps the compiler could be smart enough to autocast the literal. This already works on method calls:
```crystal
def foo(options : Hash(Symbol, Bool | String))
end

foo({:foo => true})
```

**Comment by zw963** [comment]
> I see no iconsistency. `{:foo => true}` is a hash literal and has the type `Hash(Symbol, Bool)`. You cannot assign it to a variable which restricted to the type `Hash(Symbol, Bool | String)`. If you want to do that, you can specify the concrete type with the `of` notation as in the first example:

Do we have document in crystal-book to mention this?

I propose to add those description into document, then close this issue, what about you?

**Comment by zw963** [comment]
Or add to https://github.com/crystal-lang/crystal/wiki/FAQ ?

I consider this question is representative. I have been using Crystal for half a year now, but I didn't understood it until today.

**Comment by asterite** [comment]
> I consider this question is representative

Right. It's asked at least once or twice a month. It should probably go to the FAQ explaining why sometimes `[a]` doesn't work and you have to use `[a] of A` to make it work.
